File: I1_MVPAndRetrofit/app/src/main/java/com/example/tsvetelinastoyanova/weatherreportrevisited/WeatherDetailsActivity.java
package com.example.tsvetelinastoyanova.weatherreportrevisited;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;

import com.example.tsvetelinastoyanova.weatherreportrevisited.weatherdetails.WeatherDetailsFragment;
import com.example.tsvetelinastoyanova.weatherreportrevisited.model.WeatherObject;

public class WeatherDetailsActivity extends AppCompatActivity {
    WeatherObject weatherObject;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_weather_details);
        Intent intent = getIntent();
        if (savedInstanceState == null) {
            WeatherDetailsFragment newFragment = new WeatherDetailsFragment();
            getSupportFragmentManager().beginTransaction().add(R.id.fragment_container, newFragment).commit();
            weatherObject = intent.getParcelableExtra(Constants.WEATHER_OBJECTS);
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        if (weatherObject != null) {
            WeatherDetailsFragment details = (WeatherDetailsFragment) getSupportFragmentManager().findFragmentById(R.id.fragment_container);
            details.changeView(weatherObject);
        }
    }
}

File: I1_MVPAndRetrofit_Kotlin/app/src/main/java/com/example/tsvetelinastoyanova/weatherapp/WeatherDetailsActivity.kt
package com.example.tsvetelinastoyanova.weatherapp

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import com.example.tsvetelinastoyanova.weatherapp.model.currentweather.CurrentWeatherObject
import com.example.tsvetelinastoyanova.weatherapp.util.Utils
import com.example.tsvetelinastoyanova.weatherapp.weatherdetails.WeatherDetailsContainerFragment
import com.example.tsvetelinastoyanova.weatherapp.weatherdetails.WeatherDetailsContainerPresenter

class WeatherDetailsActivity : AppCompatActivity() {
    private var weatherObject: CurrentWeatherObject? = null
    private lateinit var weatherDetailsContainerFragment: WeatherDetailsContainerFragment

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_weather_details)

        initWeatherObject()
        setupContainer()
    }

    private fun initWeatherObject() {
        if (this.weatherObject == null) {
            val weatherObject: CurrentWeatherObject? = intent.getParcelableExtra(Constants.WEATHER_OBJECTS)
            this.weatherObject = weatherObject
        }
    }

    private fun setupContainer() {
        val tempFragment: WeatherDetailsContainerFragment? =
            supportFragmentManager.findFragmentById(R.id.fragment_container) as? WeatherDetailsContainerFragment
        if (tempFragment == null) {
            weatherDetailsContainerFragment = WeatherDetailsContainerFragment.newInstance()
            Utils.addFragmentToActivity(supportFragmentManager, weatherDetailsContainerFragment, R.id.fragment_container)
        } else {
            weatherDetailsContainerFragment = tempFragment
        }
        setupContainerDetails()
    }

    private fun setupContainerDetails() {
        val weatherDetailsContainerPresenter = WeatherDetailsContainerPresenter(weatherDetailsContainerFragment,
            Utils.provideCityRepository(this))

        weatherDetailsContainerFragment.setPresenter(weatherDetailsContainerPresenter)
        weatherObject?.let {
            weatherDetailsContainerFragment.clickOnCity(it, false)
        }
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        outState?.run { putParcelable(Constants.WEATHER_OBJECTS, weatherObject) }
        super.onSaveInstanceState(outState)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle?) {
        weatherObject = savedInstanceState?.getParcelable(Constants.WEATHER_OBJECTS)
    }
}